File: domusviewer/app/src/main/java/it/achdjian/paolo/domusviewer/DomusEngineRest/WhoAreYou.java
package it.achdjian.paolo.domusviewer.DomusEngineRest;

import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Message;
import android.util.Log;

/**
 * Created by Paolo Achdjian on 14/04/16.
 */
public class WhoAreYou extends DomusEngineRest {
    private static final String TAG = WhoAreYou.class.getName();
    private static final String EXPECTED_NAME = "I am DomusEngine";
    private static Boolean running = new Boolean(false);

    public WhoAreYou(SharedPreferences sharedPreferences, ConnectionStatus connected) {
        super(sharedPreferences, connected);
    }

    public static boolean isRunning(){
        return  running;
    }

    @Override
    public void run() {
        running=true;
        try {
            String body = get("/who_are_you");
            if (body != null) {
                if (body.substring(0, EXPECTED_NAME.length()).equals(EXPECTED_NAME)) {
                    connected.setConnected(true);
                    Log.d(TAG, "Server found");
                    return;
                }
            }
            connected.setConnected(false);
            Handler handler = new Handler();
            handler.postDelayed(new WhoAreYou(this.sharedPreferences, this.connected), 1000);
        } finally {
            running=false;
        }
    }
}

File: temperature_monitor/app/src/main/java/it/achdjian/paolo/temperaturemonitor/domusEngine/rest/WhoAreYou.kt
package it.achdjian.paolo.temperaturemonitor.domusEngine.rest

import android.os.Handler
import android.util.Log
import it.achdjian.paolo.temperaturemonitor.domusEngine.ConnectionObserver
import it.achdjian.paolo.temperaturemonitor.domusEngine.ConnectionStatus
import it.achdjian.paolo.temperaturemonitor.domusEngine.DomusEngine
import it.achdjian.paolo.temperaturemonitor.domusEngine.MessageType
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Paolo Achdjian on 14/04/16.
 */
class WhoAreYou @Inject constructor(
        val domusEngine: DomusEngine,
        val domusEngineRest: DomusEngineRest,
        val connectionStatus: ConnectionStatus) : ZigbeeRunnable(){
    companion object {
        private val EXPECTED_NAME = "I am DomusEngine"
    }


    override fun run() {
        try {
            Log.i(TAG,"Start who are you")
            val body = domusEngineRest.get("/who_are_you")

            if (body.isNotBlank()) {
                if (body.substring(0, EXPECTED_NAME.length) == EXPECTED_NAME) {
                    connectionStatus.connected = true
                    Log.i(TAG,"Domus Engine found")
                    reschedule(300000)
                } else {
                    connectionStatus.connected = false
                    Log.i(TAG,"Domus Engine NOT found")
                    reschedule(10000)
                }
            } else {
                domusEngineRest.connected.connected = false
                reschedule(10000)
            }
        } catch (e: Exception) {
            domusEngineRest.connected.connected = false
            Log.i(TAG,"Domus Engine NOT found")
            reschedule(10000)
        }
    }

    private fun reschedule(delay: Long){

        domusEngine.handler.removeMessages(MessageType.WHO_ARE_YOU)
        val message = domusEngine.handler.obtainMessage(MessageType.WHO_ARE_YOU)
        if (delay==0L){
            domusEngine.handler.sendMessage(message)
        } else {
            domusEngine.handler.sendMessageDelayed(message, delay)
        }
    }

}
